src/nbl/asset/utils/ISPIRVDebloater.cpp

```diff
@@ -8,23 +8,92 @@
 
 using namespace nbl::asset;
 
+static constexpr spv_target_env SPIRV_VERSION = spv_target_env::SPV_ENV_UNIVERSAL_1_6;
+
 ISPIRVDebloater::ISPIRVDebloater()
 {
     constexpr auto optimizationPasses = std::array{
+        ISPIRVOptimizer::EOP_DEAD_BRANCH_ELIM,
+        ISPIRVOptimizer::EOP_ELIM_DEAD_FUNCTIONS,
+        ISPIRVOptimizer::EOP_DEAD_BRANCH_ELIM,
         ISPIRVOptimizer::EOP_ELIM_DEAD_FUNCTIONS,
         ISPIRVOptimizer::EOP_ELIM_DEAD_VARIABLES,
         ISPIRVOptimizer::EOP_ELIM_DEAD_CONSTANTS,
-        // TODO: remove unused type. no dedicated flag for type. maybe need agressive dce to remove type.
+        ISPIRVOptimizer::EOP_ELIM_DEAD_MEMBERS,
+        ISPIRVOptimizer::EOP_AGGRESSIVE_DCE,
         ISPIRVOptimizer::EOP_TRIM_CAPABILITIES,
     };
     m_optimizer = core::make_smart_refctd_ptr<ISPIRVOptimizer>(std::span(optimizationPasses));
 }
 
+static bool validate(const uint32_t* binary, uint32_t binarySize, nbl::system::logger_opt_ptr logger)
+{
+    auto msgConsumer = [&logger](spv_message_level_t level, const char* src, const spv_position_t& pos, const char* msg)
+    {
+        using namespace std::string_literals;
+
+
+        constexpr static nbl::system::ILogger::E_LOG_LEVEL lvl2lvl[6]{
+            nbl::system::ILogger::ELL_ERROR,
+            nbl::system::ILogger::ELL_ERROR,
+            nbl::system::ILogger::ELL_ERROR,
+            nbl::system::ILogger::ELL_WARNING,
+            nbl::system::ILogger::ELL_INFO,
+            nbl::system::ILogger::ELL_DEBUG
+        };
+        const auto lvl = lvl2lvl[level];
+        std::string location;
+        if (src)
+            location = src + ":"s + std::to_string(pos.line) + ":" + std::to_string(pos.column);
+        else
+            location = "";
+
+        logger.log(location, lvl, msg);
+    };
+    spvtools::SpirvTools core(SPIRV_VERSION);
+    core.SetMessageConsumer(msgConsumer);
+    return core.Validate(binary, binarySize);
+}
+
 ISPIRVDebloater::Result ISPIRVDebloater::debloat(const  ICPUBuffer* spirvBuffer, std::span<const EntryPoint> entryPoints, system::logger_opt_ptr logger) const
 {
     const auto* spirv = static_cast<const uint32_t*>(spirvBuffer->getPointer());
     const auto spirvDwordCount = spirvBuffer->getSize() / 4;
 
+    if (entryPoints.empty())
+    {
+       logger.log("Cannot retain zero multiple entry points!", system::ILogger::ELL_ERROR);
+       return Result{
+          nullptr,
+          false
+       };
+    }
+
+    // We will remove found entry point one by one. We set all entry points as unfound initially
+    core::set<EntryPoint> unfoundEntryPoints;
+    for (const auto& entryPoint : entryPoints)
+    {
+       if (unfoundEntryPoints.find(entryPoint) != unfoundEntryPoints.end())
```

Comment:
this should be fine, as long as the data match 100% (its just a duplicate entry in a set), only problem is if name matches but stage doesn't

Reply:
ok I asked you to use a set or map instead of span, so there will be no duplicates I guess